### PR TITLE
BUG: Reindex casting to object in certain situation

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Stumbled across this when doing #48185 
Longterm we could support a mask in take_2d_multi

### CODE REVIEW

**Comment by phofl** [comment]
cc @mroeschke gentle ping

Ci is green, forgot to remove 1.5 whatsnew

File: pandas/core/generic.py [hunk 1]
```diff
@@ -5345,6 +5345,9 @@ def _needs_reindex_multi(self, axes, method, level) -> bool_t:
             and method is None
             and level is None
             and not self._is_mixed_type
+            and not (
+                len(self.dtypes) == 1 and is_extension_array_dtype(self.dtypes.iloc[0])
```
Comment (mroeschke) [inline pandas/core/generic.py:5349]: Might be cleaner to define for `DataFrame` only? Although `Series` overrides `_needs_reindex_multi` to be `False`, `self.dtypes` for `Series` is a scalar

Comment (phofl) [reply pandas/core/generic.py]: added a ndim check

File: doc/source/whatsnew/v1.6.0.rst [hunk 2]
```diff
@@ -152,6 +152,7 @@ Interval
 Indexing
 ^^^^^^^^
 - Bug in :meth:`DataFrame.reindex` filling with wrong values when indexing columns and index for ``uint`` dtypes (:issue:`48184`)
+- Bug in :meth:`DataFrame.reindex` casting dtype to ``object`` when :class:`DataFrame` has single extension array column when indexing columns and index (:issue:`48190`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.6.0.rst:155]:
```suggestion
- Bug in :meth:`DataFrame.reindex` casting dtype to ``object`` when :class:`DataFrame` has single extension array column when re-indexing ``columns`` and ``index`` (:issue:`48190`)
```

Nit

**Review by mroeschke — APPROVED** [review]
One minor suggestion, otherwise LGTM. Feel free to merge on greenish

**Pull request merged**